spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/r2dbc/R2dbcProxyAutoConfiguration.java

```diff
@@ -0,0 +1,88 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.r2dbc;
+
+import io.r2dbc.proxy.ProxyConnectionFactory;
+import io.r2dbc.proxy.callback.ProxyConfig;
+import io.r2dbc.proxy.listener.LifeCycleExecutionListener;
+import io.r2dbc.proxy.listener.LifeCycleListener;
+import io.r2dbc.proxy.listener.ProxyExecutionListener;
+import io.r2dbc.spi.ConnectionFactory;
+
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.AnyNestedCondition;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnSingleCandidate;
+import org.springframework.boot.autoconfigure.data.r2dbc.R2dbcDataAutoConfiguration;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for R2DBC Proxy.
+ *
+ * @author Tadaya Tsuyukubo
+ * @since 2.3.1
+ */
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass(ProxyConnectionFactory.class)
+@AutoConfigureBefore(R2dbcDataAutoConfiguration.class)
+@AutoConfigureAfter(R2dbcAutoConfiguration.class)
+public class R2dbcProxyAutoConfiguration {
+
+	@Bean
+	@Primary
+	@ConditionalOnSingleCandidate(ConnectionFactory.class)
+	@Conditional(ProxyConnectionFactoryCondition.class)
+	@ConditionalOnProperty(prefix = "spring.r2dbc.proxy", value = "enabled", havingValue = "true",
+			matchIfMissing = true)
+	ConnectionFactory proxyConnectionFactory(ConnectionFactory connectionFactory,
+			ObjectProvider<ProxyConfig> proxyConfigProvider,
+			ObjectProvider<ProxyExecutionListener> proxyExecutionListeners,
+			ObjectProvider<LifeCycleListener> lifeCycleListeners) {
+		ProxyConfig proxyConfig = proxyConfigProvider.getIfAvailable(ProxyConfig::new);
+		proxyExecutionListeners.orderedStream().forEach(proxyConfig::addListener);
+		lifeCycleListeners.orderedStream()
+				.forEach((listener) -> proxyConfig.addListener(LifeCycleExecutionListener.of(listener)));
+		return ProxyConnectionFactory.builder(connectionFactory, proxyConfig).build();
+	}
+
+	static final class ProxyConnectionFactoryCondition extends AnyNestedCondition {
```

Comment:
This condition essentially states that we want to create a proxy if at least a relevant listener is around. That sounds like something that we should do in `ConnectionFactoryBuilder`. 

The fact that we have to return a different instance is a bit unusual for a customizer but something that "post-processes" the `ConnectionFactory` to create the proxy is what I have in mind. We want to make sure we only post-process the bean that has been created by auto-configuration so using a `BeanPostProcessor` might not be the best option. Modifying `ConnectionFactoryConfigurations#createConnectionFactory` could be an option.